internal/codespaces/api/api.go

```diff
@@ -188,36 +189,54 @@ type CodespaceEnvironmentConnection struct {
 
 // ListCodespaces returns a list of codespaces for the user.
 func (a *API) ListCodespaces(ctx context.Context) ([]*Codespace, error) {
-	req, err := http.NewRequest(
-		http.MethodGet, a.githubAPI+"/user/codespaces", nil,
-	)
-	if err != nil {
-		return nil, fmt.Errorf("error creating request: %w", err)
-	}
+	page := 1
+	per_page := 50
+	codespaces := []*Codespace{}
 
-	a.setHeaders(req)
-	resp, err := a.do(ctx, req, "/user/codespaces")
-	if err != nil {
-		return nil, fmt.Errorf("error making request: %w", err)
-	}
-	defer resp.Body.Close()
+	for {
+		req, err := http.NewRequest(
+			http.MethodGet, a.githubAPI+"/user/codespaces", nil,
+		)
+		if err != nil {
+			return nil, fmt.Errorf("error creating request: %w", err)
+		}
+		a.setHeaders(req)
+		q := req.URL.Query()
+		q.Add("page", strconv.Itoa(page))
+		q.Add("per_page", strconv.Itoa(per_page))
 
-	b, err := ioutil.ReadAll(resp.Body)
-	if err != nil {
-		return nil, fmt.Errorf("error reading response body: %w", err)
-	}
+		req.URL.RawQuery = q.Encode()
+		resp, err := a.do(ctx, req, "/user/codespaces")
+		if err != nil {
+			return nil, fmt.Errorf("error making request: %w", err)
+		}
+		defer resp.Body.Close()
 
-	if resp.StatusCode != http.StatusOK {
-		return nil, jsonErrorResponse(b)
-	}
+		b, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			return nil, fmt.Errorf("error reading response body: %w", err)
+		}
 
-	var response struct {
-		Codespaces []*Codespace `json:"codespaces"`
-	}
-	if err := json.Unmarshal(b, &response); err != nil {
-		return nil, fmt.Errorf("error unmarshaling response: %w", err)
+		if resp.StatusCode != http.StatusOK {
+			return nil, jsonErrorResponse(b)
+		}
+
+		var response struct {
+			Codespaces []*Codespace `json:"codespaces"`
+		}
```

Comment:
Does the server return the total? Should we capture that and use that our breaking condition instead? It feels like the current logic will do an extra round-trip to get the empty list of codespaces when we can perhaps deduce that we don't need to call again from the total count.

Reply:
Yes, we do have total count, I added it to the fake api in the test. Let me update the condition, I'll have to update the tests to catch mid-flight additions or deletions.